### PR TITLE
feat: add BuiltPayload associated type to EngineTypes

### LINKED ISSUE #6026: Introduce EngineTypes::BuiltPayload
### Describe the feature

Now that we have the Enginetypes trait, we can also make the BuiltPayload type configurable.

## TODO
* Add this as an additional associated type to `EngineTypes`
* Add associated type to `PayloadJob`

For the service this means:
It should be able to support any PayloadJob(Generator)
where `Job::Payload: Into<EngineT::BuiltPayload>` 
naming tbd

cc @loocapro

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Rjected**
Adds a `BuiltPayload` associated type and trait to `EngineTypes`

closes #6026

### CODE REVIEW

File: crates/node-api/src/engine/traits.rs [hunk 1]
```diff
@@ -2,12 +2,38 @@ use crate::{validate_version_specific_fields, AttributesValidationError, EngineA
...
+    fn fees(&self) -> U256;
+
+    /// Converts the type into the response expected by `engine_getPayloadV1`
+    fn into_v1_payload(self) -> ExecutionPayloadV1;
+
+    /// Converts the type into the response expected by `engine_getPayloadV2`
+    fn into_v2_payload(self) -> ExecutionPayloadEnvelopeV2;
+
+    /// Converts the type into the response expected by `engine_getPayloadV2`
+    fn into_v3_payload(self) -> ExecutionPayloadEnvelopeV3;
```
Comment (mattsse) [inline crates/node-api/src/engine/traits.rs:34]: these are fine for now, but perhaps we want to reconsider this eventually and instead just require that the type is Serialize or can be converted into serde_Value

File: crates/payload/builder/src/service.rs [hunk 2]
```diff
@@ -398,24 +415,24 @@ where
 }
 
 // TODO: make generic over built payload type
-type PayloadFuture =
-    Pin<Box<dyn Future<Output = Result<Arc<BuiltPayload>, PayloadBuilderError>> + Send + Sync>>;
+type PayloadFuture<P> = Pin<Box<dyn Future<Output = Result<P, PayloadBuilderError>> + Send + Sync>>;
 
 /// Message type for the [PayloadBuilderService].
-pub enum PayloadServiceCommand<T> {
+pub enum PayloadServiceCommand<A, P> {
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:421]: can we use `EngineT: EngineTypes` here?

File: crates/payload/builder/src/service.rs [hunk 3]
```diff
@@ -200,7 +208,10 @@ impl<Gen, St, Engine> PayloadBuilderService<Gen, St, Engine>
 where
     Engine: EngineTypes,
     Gen: PayloadJobGenerator,
-    Gen::Job: PayloadJob<PayloadAttributes = Engine::PayloadBuilderAttributes>,
+    Gen::Job: PayloadJob<
+        PayloadAttributes = Engine::PayloadBuilderAttributes,
+        BuiltPayload = Engine::BuiltPayload,
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:213]: I think we want:

```rust
<Gen::Job as PayloadJob>: Into<Engine::BuiltPayload>
```

this way the service can support any generator that creates jobs with a Payload that can be converted to the configured Engine::BuiltPayload.

I realize that this breaks the Arc wrapper, so I think we should remove the Arc from the Job trait function signature.

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, this is almost exactly what we want

File: crates/payload/builder/src/service.rs (hunk 2 above)
Comment (Rjected) [reply crates/payload/builder/src/service.rs]: yeah we can, just added

File: crates/payload/builder/src/service.rs (hunk 3 above)
Comment (Rjected) [reply crates/payload/builder/src/service.rs]: alright, I had to get rid of pretty much all of the Arc, might be possible to make it work again with `best_payload`, but the bounds are now `Into<Engine::BuiltPayload>`

File: crates/node-api/src/engine/traits.rs [hunk 4]
```diff
@@ -2,12 +2,38 @@ use crate::{validate_version_specific_fields, AttributesValidationError, EngineA
...
+        PayloadAttributes as EthPayloadAttributes, PayloadId, Withdrawal,
+    },
+    ExecutionPayloadV1,
 };
 
+/// Represents a built payload type that contains a built [SealedBlock] and can be converted into
+/// engine API execution payloads.
+pub trait BuiltPayload: Send + Sync + std::fmt::Debug {
+    /// Initializes the payload with the given initial block.
+    fn new(id: PayloadId, block: SealedBlock, fees: U256) -> Self;
```
Comment (mattsse) [inline crates/node-api/src/engine/traits.rs:19]: hmm, I think this is fine now, but this effectively restricts how the built payload looks like.

I see that we're using this as the fallback for empty payloads, which is a default.

let's track this separately, so we can improve this a bit.
perhaps it should be the job's responsibility to create an empty payload

File: crates/rpc/rpc-engine-api/src/engine_api.rs [hunk 5]
```diff
@@ -241,7 +241,7 @@ where
             .resolve(payload_id)
             .await
             .ok_or(EngineApiError::UnknownPayload)?
-            .map(|payload| (*payload).clone().into_v2_payload())?)
+            .map(|payload| payload.clone().into_v2_payload())?)
```
Comment (mattsse) [inline crates/rpc/rpc-engine-api/src/engine_api.rs:244]: hmm, are these clones even required here?

File: crates/rpc/rpc-engine-api/src/engine_api.rs [hunk 6]
```diff
@@ -272,7 +272,7 @@ where
             .resolve(payload_id)
             .await
             .ok_or(EngineApiError::UnknownPayload)?
-            .map(|payload| (*payload).clone().into_v3_payload())?)
+            .map(|payload| payload.clone().into_v3_payload())?)
```
Comment (mattsse) [inline crates/rpc/rpc-engine-api/src/engine_api.rs:275]: same q

**Review by mattsse — APPROVED** [review]
this is great!

last nits re clone calls

File: crates/rpc/rpc-engine-api/src/engine_api.rs (hunk 5 above)
Comment (Rjected) [reply crates/rpc/rpc-engine-api/src/engine_api.rs]: they aren't, removed

**Pull request merged**